### PR TITLE
Add LDAP_QUERY_FILTER_SENDERS setting for spoof protection with LDAP

### LINKED ISSUE #1340: Cannot send emails with SPOOF_PROTECTION=1
## Context

I have an OpenLDAP server ([osixia/openldap](https://hub.docker.com/r/osixia/openldap/) with the [postfix-book.schema](https://github.com/variablenix/ldap-mail-schema/blob/master/postfix-book.schema) added manually - seems like the version from your repository is [a bit different](https://www.diffchecker.com/tYsK8ILf), but this shouldn't change anything regarding this issue), with a user called `uid=momarqua` and `mail=moritz.marquardt@obelix.example.org`. The filters in `docker-compose.yml` are set as follows:

```yaml
LDAP_QUERY_FILTER_USER: (&(mail=%s)(mailEnabled=TRUE))
LDAP_QUERY_FILTER_GROUP: (&(mailGroupMember=%s)(mailEnabled=TRUE))
LDAP_QUERY_FILTER_ALIAS: (&(mailAlias=%s)(mailEnabled=TRUE))
LDAP_QUERY_FILTER_DOMAIN: (&(|(mail=*@%s)(mailalias=*@%s)(mailGroupMember=*@%s))(mailEnabled=TRUE))
DOVECOT_USER_FILTER: (&(objectClass=PostfixBookMailAccount)(|(mail=%u)(mailalias=%u)(uid=%n)))
DOVECOT_PASS_FILTER: (&(objectClass=PostfixBookMailAccount)(|(mail=%u)(mailalias=%u)(uid=%n)))
DOVECOT_PASS_ATTRS: uid=user,userPassword=password
SASLAUTHD_LDAP_FILTER: (&(|(mail=%U)(mailalias=%U)(uid=%U))(mailEnabled=TRUE))
```

I can receive emails without issues, I can send emails without issues, but as soon as I set `SPOOF_PROTECTION=1`, an error occurs.

## Expected Behavior
I should be able to send email as `moritz.marquardt@obelix.example.org`, or at least as any of `momarqua@obelix.example.org`, `moritz.marquardt@example.org` or `momarqua@example.org`, obviously given that I change the `mail` attribute in LDAP to the corresponding address.

## Actual Behavior

I get the following error message:

```
Dec 14 21:52:07 mail postfix/submission/smtpd[1751]: NOQUEUE: reject: RCPT from
port-1-2-3-4.dynamic.as20676.net[1.2.3.4]: 553 5.7.1 <moritz.marquardt@obelix.
example.org>: Sender address rejected: not owned by user momarqua@example.org;
from=<moritz.marquardt@obelix.example.org> to=<test@example.com> proto=ESMTP
helo=<[10.0.0.130]>
```

## Possible Fix

The workaround I'm using is to just disable `SPOOF_PROTECTION`, which obviously isn't a fix at all.

## Steps to Reproduce

TBD

## Your Environment

* Amount of RAM available: 24 GB (~20 GB free)
* Mailserver version used: `tvial/docker-mailserver:latest` pulled on 2019-12-15
* Docker version used: `18.09.6, build 481bc77`
* Environment settings relevant to the config: see **Context** above
* Any relevant stack traces ("Full trace" preferred): see **Actual Behaviour** above
* Also see #892 for a similar issue from 2018

### PR DESCRIPTION
**Pull request by moqmar**
# Description

This adds an environment variable `LDAP_QUERY_FILTER_SENDERS` that determines if a user is allowed to use a specific identity to send an email - for example, we're using something like `(|(mail=%s)(mail=noreply@example.com))`, to allow users to send an email if they either have the mail address they want to send as or if they have the mail address noreply@example.com. This applies only if SPOOF_PROTECTION=1.

~~Also, as suggested in #1340, this sets smtpd_sasl_local_domain to an empty string - it seems like without that it won't work, but I couldn't figure out the implications of that, so that might be worth discussing.~~ (that change has been reverted as the issue itself seems to have been fixed in the meantime)

If the environment variable is empty, the old sender `smtpd_sender_login_maps` behaviour will be used.

Fixes #1340

## Type of change

- [ ] ~~Bug fix (non-breaking change which fixes an issue)~~
- [x] New feature (non-breaking change which adds functionality)

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation (README.md or ENVIRONMENT.md or the documentation)
- [ ] If necessary I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

**Comment by aendeavor** [comment]
LGTM, but I'm not using LDAP myself. I will request a review.

**Comment by reneploetz** [comment]
Setting  smtpd_sasl_local_domain to an empty string changes the lookup key for the authentication in case the SMTP client does not send an authentication with a domain name, e.g. if he tries to authenticate with "foo" and smtpd_sasl_local_domain = "bar.com", then he would authenticate with "foo@bar.com" against SASL.
I *think* this might break existing configurations if the SASLAUTHD_LDAP_FILTER is used without a domain name.

Note that SPOOF_PROTECTION=1 does already work on my system using LDAP aliases for my accounts (*), so I *think* this is more a configuration issue, especially given that you replace all other lookups with a new lookup. Removing the requirement for $mydomain in smtpd_sasl_local_domain leads me to believe that the underlying problem is that your initial lookup maps do not match your domain settings.
Given your original bug report, can you try to use just the UID for the lookup in the SASLAUTHD_LDAP_FILTER?


(*) I'm currently using LDAP with a Kopano-LDAP-schema like so:
```
LDAP_QUERY_FILTER_USER=(&(kopanoAccount=1)(mail=%s))
LDAP_QUERY_FILTER_GROUP=(&(objectclass=kopano-group)(mail=%s))
LDAP_QUERY_FILTER_ALIAS=(&(kopanoAccount=1)(kopanoAliases=%s))
LDAP_QUERY_FILTER_DOMAIN=(mail=*@%s)
SASLAUTHD_LDAP_FILTER=(&(kopanoAccount=1)(uid=%s))
```

Both mail and kopanoAliases attributes contain fully qualified email addresses. The uid is the local part without the domain name.

**Comment by moqmar** [comment]
Ah, your SASLAUTHD_LDAP_FILTER isn't being used at all because you're using rimap. According to https://blog.sys4.de/cyrus-sasl-saslauthdconf-man-page-en.html, `%s` is the service name: in my case it looks up `(uid=smtp)` if I set it that way. `%D` is empty; `%u` is just the part before the @, `%d` is always empty for some reason, even when specifying a domain. `(mail=%u@example.org)` is the only thing that works, but then `mail@blah.com` is treated the same as `mail@example.org`, although that error can then be caught later by the spoof protection when using the email address as an identifier.

When using that `(mail=%u@example.org)` filter, you're right that spoof protection seems to work correctly now, so it seems like the original #1340 issue has solved itself or always was a configuration issue?

**TL;DR: SASLAUTHD_LDAP_FILTER doesn't really work reliably in general, and has not much to do with this; when leaving `smtpd_sasl_local_domain` as it was before it seems like everything still works as intended, and #1340 seems to be solved. I have changed it back, and thanks for the `rimap` tip.**

---

So, it doesn't seem like this solves #1340 (as it seems to be working now in general), but I'd still like to see this merged. The big advantage of the sender filter (even if it works without this PR) is that it just provides a lot more flexibility. I'm using basically the following filter for example:
```
(&
  (memberOf=cn=employee,ou=groups,dc=example,dc=org)
  (|
    (&(mail=%u@example.org)(mail=%s))
    (mail=noreply@example.org)
    (memberOf=cn=admin,ou=groups,dc=example,dc=org)
  )
)
```

This means that only `employee` members can send any emails (even if others might have an IMAP inbox and can receive emails), only from their email address `@example.org` (even if they also have an additional `@blah.com` in the LDAP, and maybe even can receive emails on that address). Also, `noreply@example.org` and members of the `admin` group can send emails from any address.

That kind of flexibility would be impossible to achieve without an additional sender filter table.

**Review by wernerfred — COMMENTED** [review]
LGTM 👍🏻 
I just want someone else to look over it before approving this

**Review by wernerfred — CHANGES_REQUESTED** [review]
Can you please write another test that shows that your changes work both ways?

This should be the correct file:
https://github.com/docker-mailserver/docker-mailserver/blob/master/test/mail_with_ldap.bats

**Comment by moqmar** [comment]
Test has been added - I set the variable so that additionally to the original behaviour, the user `some.user.id` (from "04_user-email-different-uid.ldif") can use any address for sending.

The "rejects sender forging" test then checks if spoofing fails for `some.user` (from "02_user-email.ldif"), and the "uses senders filter" checks if `some.user.id` *can* spoof its address due to the exception in the filter.

**Review by wernerfred — APPROVED** [review]
Thanks for your changes. Really appreciate your contribution and the fact that you wrote some tests!

I will approve this PR **BUT** I would really like to see the nice explanation you already provided in the documentation.

Can you add this:

>The big advantage of the sender filter (even if it works without this PR) is that it just provides 
>
> a lot more flexibility. I'm using basically the following filter for example:
> 
> (&
>   (memberOf=cn=employee,ou=groups,dc=example,dc=org)
>   (|
>     (&(mail=%u@example.org)(mail=%s))
>     (mail=noreply@example.org)
>     (memberOf=cn=admin,ou=groups,dc=example,dc=org)
>   )
> )
> This means that only employee members can send any emails (even if others might have an IMAP inbox and can receive emails), only from their email address @example.org (even if they also have an additional @blah.com in the LDAP, and maybe even can receive emails on that address). Also, noreply@example.org and members of the admin group can send emails from any address.
> 
> That kind of flexibility would be impossible to achieve without an additional sender filter table.

to [probably this page](https://docker-mailserver.github.io/docker-mailserver/edge/config/advanced/auth-ldap/) (or another one if you find a better spot) before this gets merged?

**Review by wernerfred — CHANGES_REQUESTED** [review]
Okay after finishing my review i realized that there are more changes so i will swith to request changes but in general i like your changes and will approve based on your next action ;)

File: target/postfix/ldap-senders.cf [hunk 1]
```diff
@@ -0,0 +1,9 @@
+bind             = yes
+bind_dn          = cn=admin,dc=domain,dc=com
+bind_pw          = admin
+query_filter     = (mail=%s)
+result_attribute = mail, uid
+search_base      = ou=people,dc=domain,dc=com
+server_host      = mail.domain.com
```
[the remark this answers is not in the document]
Comment (aendeavor) [reply target/postfix/ldap-senders.cf:7]: Rebasing will show. I'd definiely merge #1901 first.

File: test/mail_with_ldap.bats [hunk 2]
```diff
@@ -191,11 +192,15 @@ function teardown_file() {
   assert_success
 }
 
-# ATTENTION: this test must come after "checking dovecot: ldap mail delivery works" since it will deliver an email which skews the count in said test, leading to failure
+# ATTENTION: these tests must come after "checking dovecot: ldap mail delivery works" since they will deliver an email which skews the count in said test, leading to failure
```
[the remark this answers is not in the document]
Comment (moqmar) [reply test/mail_with_ldap.bats:195]: Only if it fails because the email could be sent (even if it shouldn't) - if it passes, no email was sent.

File: target/postfix/ldap-senders.cf (hunk 1 above)
[the remark this answers is not in the document]
Comment (moqmar) [reply target/postfix/ldap-senders.cf:7]: Nope, postfix already allows `server_host` to be `ldaps://hostname` (see http://www.postfix.org/ldap_table.5.html); #1901 is purely a Dovecot issue.

**Comment by aendeavor** [comment]
Please also note that your commit signature is invalid.

**Comment by moqmar** [comment]
For the LDAP documentation, I'd like to generally create a separate PR as the other variables aren't really explained yet either.

**Comment by wernerfred** [comment]
> For the LDAP documentation, I'd like to generally create a separate PR as the other variables aren't really explained yet either, is that fine for you?

Absolutely fine. Would be wonderful if you could give the LDAP documentation some love. Feel free to open the PR as WIP so we can keep track of ongoing changes and support you :)

**Review by wernerfred — APPROVED** [review]
@aendeavor wanna still wait for #1901 ? ;)  I let merging this up to you. 

In #1901 I want at least another one with LDAP experience to review it.

**Comment by aendeavor** [comment]
> @aendeavor wanna still wait for #1901 ? ;)  I let merging this up to you. 

Fine with me, there's no need to wait :D I will merge this.

> In #1901 I want at least another one with LDAP experience to review it.

Yes, that's a good idea. I don't have experience with LDAP either.
